src/rail_gui.cpp

```diff
@@ -762,6 +762,45 @@ struct BuildRailToolbarWindow : Window {
 	static HotkeyList hotkeys;
 };
 
+/**
+ * Selects RailType based on _last_built_railtype and global modifiers from RailToolbarWidgetModifiers
+ *
+ * If GHK_MOD_* modifiers were used hotkey replaced with last selected widget
+ * or WID_RAT_AUTORAIL (if toolbar was closed).
+ *
+ * Without wrapping, so if you're on RAILTYPE_RAIL you'll stay there. Same for upper boundary.
+ *
+ * @param [out] hotkey Replaced by last user selected widget or WID_RAT_AUTORAIL if prev/next railtype used
+ * @return selected RailType
+ */
+static RailType ChooseRailTypeOnGlobalHotkey(int &hotkey)
+{
+	extern RailType _last_built_railtype;
+	RailType rt = _last_built_railtype;
+	bool change_rt = false;
+	if (HasBit(hotkey, GHK_MOD_RAT_PREV_RAILTYPE)) {
+		ClrBit(hotkey, GHK_MOD_RAT_PREV_RAILTYPE);
+		change_rt = true;
+		do {
+			rt--;
+		} while (rt > RAILTYPE_BEGIN && !HasRailtypeAvail(_local_company, rt));
+	} else if (HasBit(hotkey, GHK_MOD_RAT_NEXT_RAILTYPE)) {
+		ClrBit(hotkey, GHK_MOD_RAT_NEXT_RAILTYPE);
+		change_rt = true;
+		do {
+			rt++;
+		} while (rt < RAILTYPE_END && !HasRailtypeAvail(_local_company, rt));
+		if (rt >= RAILTYPE_END) rt = _last_built_railtype;
+	}
+	if (change_rt) {
+		// get previous selected widget from current build toolbar if present
+		auto *w = dynamic_cast<BuildRailToolbarWindow *>(FindWindowById(WC_BUILD_TOOLBAR, TRANSPORT_RAIL));
+		hotkey = (w == nullptr || w->last_user_action < WID_RAT_BUILD_NS) ? WID_RAT_AUTORAIL : w->last_user_action;
```

Comment:
You're not copying the bulldozer button state (WID_RAT_REMOVE).

Reply:
Oh, and also it selects autoraill if no widgets were previously selected.

Reply:
> You're not copying the bulldozer button state (WID_RAT_REMOVE).

Could you elaborate on that? It will use `w->last_user_action` for any `WID_RAT_*` except `WID_RAT_CAPTION`.

And selection autorail was by design, however I can remove it to make it orthogonal. I think select autorail is useful from UX perspective.